delta-lake/src/main/java/org/apache/iceberg/delta/BaseSnapshotDeltaLakeTableAction.java

```diff
@@ -0,0 +1,370 @@
+  private TableIdentifier newTableIdentifier;
+  private String newTableLocation;
+  private HadoopFileIO deltaLakeFileIO;
+
+  /**
+   * Snapshot a delta lake table to be an iceberg table. The action will read the delta lake table's
+   * log through the table's path, create a new iceberg table using the given icebergCatalog and
+   * newTableIdentifier, and commit all changes in one iceberg transaction.
+   *
+   * <p>The new table will only be created if the snapshot is successful.
+   *
+   * @param deltaTableLocation the delta lake table's path
+   */
+  BaseSnapshotDeltaLakeTableAction(String deltaTableLocation) {
+    this.deltaTableLocation = deltaTableLocation;
+    this.newTableLocation = deltaTableLocation;
+  }
+
+  @Override
+  public SnapshotDeltaLakeTable tableProperties(Map<String, String> properties) {
+    additionalPropertiesBuilder.putAll(properties);
+    return this;
+  }
+
+  @Override
+  public SnapshotDeltaLakeTable tableProperty(String name, String value) {
+    additionalPropertiesBuilder.put(name, value);
+    return this;
+  }
+
+  @Override
+  public SnapshotDeltaLakeTable tableLocation(String location) {
+    this.newTableLocation = location;
+    return this;
+  }
+
+  @Override
+  public SnapshotDeltaLakeTable as(TableIdentifier identifier) {
+    this.newTableIdentifier = identifier;
+    return this;
+  }
+
+  @Override
+  public SnapshotDeltaLakeTable icebergCatalog(Catalog catalog) {
+    this.icebergCatalog = catalog;
+    return this;
+  }
+
+  @Override
+  public SnapshotDeltaLakeTable deltaLakeConfiguration(Configuration conf) {
+    this.deltaLog = DeltaLog.forTable(conf, deltaTableLocation);
+    this.deltaLakeFileIO = new HadoopFileIO(conf);
+    return this;
+  }
+
+  @Override
+  public Result execute() {
+    Preconditions.checkArgument(
+        icebergCatalog != null && newTableIdentifier != null,
+        "Iceberg catalog and identifier cannot be null. Make sure to configure the action with a valid Iceberg catalog and identifier.");
+    Preconditions.checkArgument(
+        deltaLog != null && deltaLakeFileIO != null,
+        "Make sure to configure the action with a valid deltaLakeConfiguration");
+    io.delta.standalone.Snapshot updatedSnapshot = deltaLog.update();
+    Schema schema = convertDeltaLakeSchema(updatedSnapshot.getMetadata().getSchema());
+    PartitionSpec partitionSpec = getPartitionSpecFromDeltaSnapshot(schema);
+    Transaction icebergTransaction =
+        icebergCatalog.newCreateTableTransaction(
+            newTableIdentifier,
+            schema,
+            partitionSpec,
+            newTableLocation,
+            destTableProperties(updatedSnapshot, deltaTableLocation));
+    icebergTransaction
+        .table()
+        .updateProperties()
+        .set(
+            TableProperties.DEFAULT_NAME_MAPPING,
+            NameMappingParser.toJson(MappingUtil.create(icebergTransaction.table().schema())))
+        .commit();
+    Iterator<VersionLog> versionLogIterator =
```

Comment:
Thank you for your suggestions. I would like to discuss about my thought on the current design:

In this implementation, my intention is to make all delta lake tables' snapshots available in the newly created iceberg table so that users can time travel to any snapshot they want after the snapshot/migration. Hence, I iterates from the beginning and make each version log as a commit to the iceberg table. 

With [`DeltaLog.snapshot()`](https://delta-io.github.io/connectors/latest/delta-standalone/api/java/io/delta/standalone/DeltaLog.html#snapshot--), we can access the latest checkpoint and build the new iceberg from here. However, it would make user lose acess to all snapshots before the last checkpoint in the new iceberg table. 

There is another option: With [`DeltaLog.update()`](https://delta-io.github.io/connectors/latest/delta-standalone/api/java/io/delta/standalone/DeltaLog.html#update--), we can access the latest snapshot (built from the latest checkpoint) and only commits the latest snapshot to the new iceberg table. This approach preserve the fewest logs but is the fastest.

In conclusion, I think the tradeoff here is between the action time cost and the number of logs converted to the new iceberg table. I feel that the later two approach can be a user-configurable option (perhaps with a name like `shallowCopy`?) so that user can choose not to convert previous snapshots if they do not need them. I think this option can be added later. 

Do you think it is ok to keep the current implementation as the initial PR for this module? Please correct me if I misunderstand something.

Reply:
I did not totally understand Prashant's question, but now I understand (if what Jonas's interpretation is correct). In that case I would +1 for having all transactions migrated. Only migrating the last transaction could just be a feature flag we add later, but I think most people would like a exact 1:1 mapping of those table formats including the transactions. 

Also this will align well with what I am thinking about adding after this PR is merged, which is a `addTransactions` method to add new transactions, similar to `add_files` procedure in Hive to Iceberg migration.

Reply:
I see, my only concern was slowness of the migration as we are iteration one version log at time which could be very well skipped if we would have directly read the checkpointed file and the version logs after.

> but I think most people would like a exact 1:1 mapping of those table formats including the transactions

I see if that's the case, then +1 from my side as well.

side note : how are we time traveling with same semantics as delta in the new iceberg table (or is it something we are planning to support), are we thinking of tagging the new iceberg snapshots with delta commit id ? so that `VERSION AS OF` still works ?

Reply:
I was proposing a `AS OF 'delta-xxxxxx'` using the syntax I just merged for querying tags #6575 , as as follow up for this PR